cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java

```diff
@@ -0,0 +1,236 @@
+      // from, and then we perform the swap. If this check fails, the thread performing the swap does not get ownership
+      // of the agent.
+      // Swap happens from KEYS[1] -> KEYS[2] with the agent type being ARGV[1], and the score being ARGV[2].
+      scriptShas.put(SWAP_SET_SCRIPT, jedis.scriptLoad(
+          "if redis.call('zrank', KEYS[1], ARGV[1]) ~= nil then\n" +
+          "  redis.call('zrem', KEYS[1], ARGV[1])\n" +
+          "  return redis.call('zadd', KEYS[2], ARGV[2], ARGV[1])\n" +
+          "else return nil end\n"
+      ));
+
+      // If the agent isn't present in either the WAITING or WORKING sets, it's safe to add. If it's present in either,
+      // it's being worked on or was recently run, so leave it be.
+      // KEYS[1] and KEYS[2] are checked for inclusion. If the agent is in neither ARGV[1] is added to KEYS[1] with score
+      // ARGV[2].
+      scriptShas.put(ADD_AGENT_SCRIPT, jedis.scriptLoad(
+        "if redis.call('zrank', KEYS[1], ARGV[1]) ~= nil then\n" +
+        "  if redis.call('zrank', KEYS[2], ARGV[1]) ~= nil then\n" +
+        "    return redis.call('zadd', KEYS[1], ARGV[2], ARGV[1])\n" +
+        "  else return nil end\n" +
+        "else return nil end\n"));
+    }
+  }
+
+  @Override
+  public void schedule(Agent agent, AgentExecution agentExecution, ExecutionInstrumentation executionInstrumentation) {
+    if (agent instanceof AgentSchedulerAware) {
+      ((AgentSchedulerAware)agent).setAgentScheduler(this);
+    }
+    agents.put(agent.getAgentType(), new AgentWorker(agent, agentExecution, executionInstrumentation, this));
+    try (Jedis jedis = jedisSource.getJedis()) {
+      jedis.evalsha(scriptShas.get(ADD_AGENT_SCRIPT), 2, WAITING_SET, WORKING_SET, agent.getAgentType(), score(NOW));
+    }
+  }
+
+  @Override
+  public void run() {
+    if (!nodeStatusProvider.isNodeEnabled()) {
+      return;
+    }
+    try {
+      saturatePool();
+    } catch (Throwable t) {
+      log.error("Failed to run caching agents", t);
+    } finally {
+      runCount++;
+    }
+  }
+
+  private static String score(long offset) {
+    return String.format("%d", (TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()) + offset));
+  }
+
+  private boolean acquireAgent(Agent agent) {
+    try (Jedis jedis = jedisSource.getJedis()) {
+      return jedis.evalsha(scriptShas.get(SWAP_SET_SCRIPT),
+          Arrays.asList(WAITING_SET, WORKING_SET),
+          Arrays.asList(agent.getAgentType(), score(intervalProvider.getInterval(agent).getTimeout()))) != null;
+    }
+  }
+
+  private boolean releaseAgent(Agent agent) {
+    try (Jedis jedis = jedisSource.getJedis()) {
+      if (jedis.evalsha(scriptShas.get(SWAP_SET_SCRIPT),
+          Arrays.asList(WORKING_SET, WAITING_SET),
+          Arrays.asList(agent.getAgentType(), score(intervalProvider.getInterval(agent).getTimeout()))) != null) {
+        runningAgents.release();
```

Comment:
NPE if parallelism = -1

maybe change runningAgents to Optional<Semaphore> and do `runningAgents.ifPresent(Semaphore::release)` here and below `runningAgents.map(Semaphore::tryAcquire).orElse(true)`